web/src/main/java/org/springframework/security/web/authentication/MultiTenantAuthenticationManagerResolver.java

```diff
@@ -0,0 +1,174 @@
+import org.springframework.web.util.UriComponents;
+import org.springframework.web.util.UriComponentsBuilder;
+
+/**
+ * An implementation of {@link AuthenticationManagerResolver} that separates the tasks of
+ * extracting the request's tenant identifier and looking up an {@link AuthenticationManager}
+ * by that tenant identifier.
+ *
+ * @author Josh Cummings
+ * @since 5.2
+ * @see AuthenticationManagerResolver
+ */
+public final class MultiTenantAuthenticationManagerResolver<T> implements AuthenticationManagerResolver<HttpServletRequest> {
+
+	private final Converter<HttpServletRequest, AuthenticationManager> authenticationManagerResolver;
+
+	/**
+	 * Constructs a {@link MultiTenantAuthenticationManagerResolver} with the provided parameters
+	 *
+	 * @param tenantResolver
+	 * @param authenticationManagerResolver
+	 */
+	public MultiTenantAuthenticationManagerResolver
+			(Converter<HttpServletRequest, T> tenantResolver,
+					Converter<T, AuthenticationManager> authenticationManagerResolver) {
+
+		Assert.notNull(tenantResolver, "tenantResolver cannot be null");
+		Assert.notNull(authenticationManagerResolver, "authenticationManagerResolver cannot be null");
+
+		this.authenticationManagerResolver = request -> {
+			Optional<T> context = Optional.ofNullable(tenantResolver.convert(request));
+			return context.map(authenticationManagerResolver::convert)
+					.orElseThrow(() -> new IllegalArgumentException
+							("Could not resolve AuthenticationManager by reference " + context.orElse(null)));
+		};
+	}
+
+	@Override
+	public AuthenticationManager resolve(HttpServletRequest context) {
+		return this.authenticationManagerResolver.convert(context);
+	}
+
+	/**
+	 * Creates an {@link AuthenticationManagerResolver} that will use a hostname's first label as
+	 * the resolution key for the underlying {@link AuthenticationManagerResolver}.
+	 *
+	 * For example, you might have a set of {@link AuthenticationManager}s defined like so:
+	 *
+	 * <pre>
+	 * 	Map<String, AuthenticationManager> authenticationManagers = new HashMap<>();
+	 *  authenticationManagers.put("tenantOne", managerOne());
+	 *  authenticationManagers.put("tenantTwo", managerTwo());
+	 * </pre>
+	 *
+	 * And that your system serves hostnames like <pre>https://tenantOne.example.org</pre>.
+	 *
+	 * Then, you could create an {@link AuthenticationManagerResolver} that uses the "tenantOne" value from
+	 * the hostname to resolve Tenant One's {@link AuthenticationManager} like so:
+	 *
+	 * <pre>
+	 *	AuthenticationManagerResolver<HttpServletRequest> resolver =
+	 *			resolveByFirstHostnameLabel(authenticationManagers::get);
+	 * </pre>
+	 *
+	 * {@link HttpServletRequest}
+	 * @param resolver A {@link String}-resolving {@link AuthenticationManagerResolver}
+	 * @return A hostname-resolving {@link AuthenticationManagerResolver}
+	 */
+	public static AuthenticationManagerResolver<HttpServletRequest>
+			resolveByFirstHostnameLabel(Converter<String, AuthenticationManager> resolver) {
```

Comment:
I'm not sure about the phrasing for this. Perhaps it is just me, but I typically hear the phrase subdomain rather than hostname label. Have you given thought to resolveBySubdomain?

Reply:
I'm all for pragmatism, and "subdomain" may be more pragmatic. I think it reads more cleanly, too. 

The reason that I avoided it is because of [Wikipedia's definition of a subdomain](https://en.wikipedia.org/wiki/Subdomain): 

> A subdomain is a domain that is part of a larger domain; the only domain that is not also a subdomain is the root domain.[1] For example, west.example.com and east.example.com/ are subdomains of the example.com domain, which in turn is a subdomain of the com top-level domain (TLD)

If that's the case, then "west" or "east" are not subdomains, even though that's what we all often call them.

Knowing this, what are your thoughts? We could call it `resolveBySubdomain` (which is what I originally had), but would it cause cognitive dissonance?

Reply:
Thanks for the comment. I understand your reservations. I guess I wouldn't necessarily assume that the value of the tenant is the subdomain itself, but that the tenant is resolved from a subdomain. Perhaps a wording of `resolveFromSubdomain` would be better?